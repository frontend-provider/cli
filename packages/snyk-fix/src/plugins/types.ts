import { EntityToFix } from '../types';

export type FixHandler = (entities: EntityToFix[]) => Promise<FixHandlerResult[]>;

export interface FixHandlerResult {
  succeeded: EntityToFix[];
  failed: EntityToFix[];
}
