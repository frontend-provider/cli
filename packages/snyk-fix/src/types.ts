import { DepGraphData } from '@snyk/dep-graph';
import { CustomError } from './lib/errors/custom-error';

/* Scan Result
 * this data is returned by the CLI plugins to identify
 * what should be scanned for issues
 */
export interface GitTarget {
  remoteUrl: string;
  branch: string;
}
export interface ContainerTarget {
  image: string;
}

export interface ScanResult {
  readonly identity: Identity;
  readonly facts: Facts[];
  readonly name?: string;
  readonly policy?: string;
  readonly target?: GitTarget | ContainerTarget;
}

export interface Identity {
  type: string;
  targetFile?: string;
  // options used to scan should be here
  args?: { [key: string]: string };
}

export interface Facts {
  type: string;
  data: any;
}

/* Test Result
 * this data is returned on a `snyk test` for supported project types
 * after the relevant plugin extracts dependencies
 */
export interface TestResult {
  readonly issues: Issue[];
  readonly issuesData: IssuesData;
  readonly depGraphData: DepGraphData;
  readonly remediation?: RemediationChanges; // TODO: not yet in the CLI TestResults type
}

export interface Issue {
  pkgName: string;
  pkgVersion?: string;
  issueId: string;
  fixInfo: FixInfo;
}

interface UpgradePath {
  path: UpgradePathItem[];
}

export interface FixInfo {
  upgradePaths?: UpgradePath[];
  isPatchable?: boolean;
  nearestFixedInVersion?: string;
}

interface UpgradePathItem {
  name: string;
  version: string;
  newVersion?: string;
  isDropped?: boolean;
}
export interface IssuesData {
  [issueId: string]: {
    id: string;
    severity: SEVERITY;
    title: string;
  };
}

/* Remediation Data
 * this data is returned on a `snyk test` for supported project types
 */
export interface Upgrade {
  upgradeTo: string; // name@version
}

export interface UpgradeVulns extends Upgrade {
  vulns: string[];
}
export interface UpgradeRemediation extends UpgradeVulns {
  upgrades: string[];
}

export interface PatchRemediation {
  paths: PatchObject[];
}

export interface DependencyUpdates {
  [from: string]: UpgradeRemediation;
}

export interface PinRemediation extends UpgradeVulns {
  isTransitive: boolean;
}

export interface DependencyPins {
  [name: string]: PinRemediation;
}

// Remediation changes to be applied to the project,
// including information on all and unresolved issues.
export interface RemediationChanges {
  unresolved: IssueData[];
  upgrade: DependencyUpdates;
  patch: {
    [name: string]: PatchRemediation;
  };
  ignore: unknown;
  pin: DependencyPins;
}

export interface IssueData {
  id: string;
  packageName: string;
  version: string;
  moduleName?: string;
  below: string; // Vulnerable below version
  semver: {
    vulnerable: string | string[];
    vulnerableHashes?: string[];
    vulnerableByDistro?: {
      [distroNameAndVersion: string]: string[];
    };
  };
  patches: Patch[];
  isNew: boolean;
  description: string;
  title: string;
  severity: SEVERITY;
  fixedIn: string[];
  legalInstructions?: string;
  reachability?: REACHABILITY;
}

interface Patch {
  version: string;
  id: string;
  urls: string[];
  modificationTime: string;
}

export enum REACHABILITY {
  FUNCTION = 'function',
  PACKAGE = 'package',
  NOT_REACHABLE = 'not-reachable',
  NO_INFO = 'no-info',
}

export interface PatchObject {
  [name: string]: {
    patched: string;
  };
}

export enum SEVERITY {
  LOW = 'low',
  MEDIUM = 'medium',
  HIGH = 'high',
  CRITICAL = 'critical',
}

/* End Remediation Data
 */

/* Snyk fix types
 * Types for concepts introduced as part of this lib
 */

export type SupportedScanTypes = 'pip';

export interface Workspace {
  path: string;
  readFile: (path: string) => Promise<string>;
  writeFile: (path: string, content: string) => Promise<void>;
}
export interface EntityToFix {
  readonly workspace: Workspace;
  readonly scanResult: ScanResult;
  readonly testResult: TestResult;
  readonly options: CliTestOptions;
}

// Partial CLI test options interface
// defining only what is used by @snyk/fix
// add more as needed
export interface PythonTestOptions {
  command?: string; // python interpreter to use for python tests
  dev?: boolean;
}
export type CliTestOptions = PythonTestOptions;
export interface WithError<Original> {
  original: Original;
  error: CustomError;
  tip?: string;
}

export interface WithFixChangesApplied<Original> {
  original: Original;
  changes: FixChangesSummary[];
}

export interface WithUserMessage<Original> {
  original: Original;
  userMessage: string;
}

export type FixChangesSummary = FixChangesSuccess | FixChangesError;

interface FixChangesSuccess {
  success: true;
  userMessage: string;
  issueIds: string[];
  from?: string;
  to?: string;
}

interface FixChangesError {
  success: false;
  userMessage: string;
  reason: string;
  tip?: string;
  issueIds: string[];
  from?: string;
  to?: string;
}

export interface ErrorsByEcoSystem {
  [ecosystem: string]: { originals: EntityToFix[]; userMessage: string };
}
export interface FixOptions {
  dryRun?: boolean;
  quiet?: boolean;
  stripAnsi?: boolean;
}

export interface FixedMeta {
  fixed: number;
  failed: number;
  fixableIssues: number;
  fixedIssues: number;
  totalIssues: number;
}
