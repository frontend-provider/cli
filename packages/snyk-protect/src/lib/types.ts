export type PatchInfo = {
  patchableVersions?: string;
  urls: string[];
};

export type VulnIdAndPackageName = {
  vulnId: string;
  packageName: string;
};

export type FoundPhysicalPackage = {
  packageName: string;
  packageVersion: string;
  path: string;
};

export type Patch = {
  patchableVersions?: string;
  patchDiffs: string[];
};

export type VulnPatches = {
  vulnId: string;
  patches: Patch[];
};

export type PatchedModule = {
  vulnId: string;
  packageName: string;
  packageVersion: string;
};

export enum ProtectResultType {
  NO_SNYK_FILE = 'NO_SNYK_FILE',
  NOTHING_TO_PATCH = 'NOTHING_TO_PATCH',
  APPLIED_PATCHES = 'APPLIED_PATCHES',
}

export type AnalyticsPayload = {
  command: string;
  args: string[];
  version: string;
  nodeVersion: string;
  metadata: {
    protectResult: ProtectResult;
  };
};

export type NoSnykFile = {
  type: ProtectResultType.NO_SNYK_FILE;
};

export type NothingToPatch = {
  type: ProtectResultType.NOTHING_TO_PATCH;
};

export type AppliedPatches = {
  type: ProtectResultType.APPLIED_PATCHES;
  patchedModules: PatchedModule[];
};

export type ProtectResult = NoSnykFile | NothingToPatch | AppliedPatches;
