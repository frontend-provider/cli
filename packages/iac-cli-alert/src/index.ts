import {
  IncomingWebhook,
  IncomingWebhookDefaultArguments,
} from '@slack/webhook';

if (!process.env.IAC_SMOKE_TESTS_SLACK_WEBHOOK_URL) {
  console.error(
    'Missing the IAC_SMOKE_TESTS_SLACK_WEBHOOK_URL environment variable',
  );
  process.exit(1);
}

if (!process.env.IAC_SMOKE_TESTS_OS) {
  console.error('Missing the IAC_SMOKE_TESTS_OS environment variable');
  process.exit(1);
}

const IAC_SMOKE_TESTS_SLACK_WEBHOOK_URL =
  process.env.IAC_SMOKE_TESTS_SLACK_WEBHOOK_URL;
const IAC_SMOKE_TESTS_OS = process.env.IAC_SMOKE_TESTS_OS;

const slackWebhook = new IncomingWebhook(IAC_SMOKE_TESTS_SLACK_WEBHOOK_URL);

async function sendSlackAlert() {
  console.log('IaC smoke tests failed. Sending Slack alert...');
  const args: IncomingWebhookDefaultArguments = {
    username: 'IaC CLI Alerts',
    text: `An Infrastructure as Code Smoke Tests job failed. \n Operating system - ${IAC_SMOKE_TESTS_OS} \n<https://github.com/snyk/cli/actions?query=workflow%3A+%22Infrastructure+as+Code+Smoke+Tests%22|Infrastructure as Code Smoke Tests Results>`,
    icon_emoji: 'snyk-iac',
  };
  await slackWebhook.send(args);
  console.log('Slack alert sent.');
}

async function run() {
  try {
    await sendSlackAlert();
  } catch (error) {
    console.error(error);
    process.exit(1);
  }
}

process.on('uncaughtException', (err) => {
  console.error(err);
  process.exit(1);
});

// Exec
run();
