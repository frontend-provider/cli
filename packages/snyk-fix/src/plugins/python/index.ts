import * as debugLib from 'debug';
import * as pMap from 'p-map';
import * as ora from 'ora';

import { EntityToFix, FixOptions } from '../../types';
import { FixHandlerResultByPlugin } from '../types';
import { loadHandler } from './load-handler';
import { SUPPORTED_HANDLER_TYPES } from './supported-handler-types';
import { mapEntitiesPerHandlerType } from './map-entities-per-handler-type';
import chalk = require('chalk');
import { partitionByFixable } from './handlers/is-supported';

const debug = debugLib('snyk-fix:python');

export async function pythonFix(
  entities: EntityToFix[],
  options: FixOptions,
): Promise<FixHandlerResultByPlugin> {
  const spinner = ora({ isSilent: options.quiet, stream: process.stdout });
  const spinnerMessage = 'Looking for supported Python items';
  spinner.text = spinnerMessage;
  spinner.start();

  const handlerResult: FixHandlerResultByPlugin = {
    python: {
      succeeded: [],
      failed: [],
      skipped: [],
    },
  };
  const results = handlerResult.python;
  const { entitiesPerType, skipped: notSupported } = mapEntitiesPerHandlerType(
    entities,
  );
  results.skipped.push(...notSupported);

  spinner.stopAndPersist({
    text: spinnerMessage,
    symbol: chalk.green('\n✔'),
  });

  await pMap(
    Object.keys(entitiesPerType),
    async (projectType) => {
      const projectsToFix: EntityToFix[] = entitiesPerType[projectType];
      if (!projectsToFix.length) {
        return;
      }
      const processingMessage = `Processing ${projectsToFix.length} ${projectType} items`;
      const processedMessage = `Processed ${projectsToFix.length} ${projectType} items`;

      spinner.text = processingMessage;
      spinner.render();

      try {
        const handler = loadHandler(projectType as SUPPORTED_HANDLER_TYPES);
        // drop unsupported Python entities early so only potentially fixable items get
        // attempted to be fixed
        const { fixable, skipped: notFixable } = await partitionByFixable(
          projectsToFix,
        );
        results.skipped.push(...notFixable);

        const { failed, skipped, succeeded } = await handler(fixable, options);
        results.failed.push(...failed);
        results.skipped.push(...skipped);
        results.succeeded.push(...succeeded);
      } catch (e) {
        debug(
          `Failed to fix ${projectsToFix.length} ${projectType} projects.\nError: ${e.message}`,
        );
        results.failed.push(
          ...projectsToFix.map((p) => ({ original: p, error: e })),
        );
      }
      spinner.stopAndPersist({
        text: processedMessage,
        symbol: chalk.green('✔'),
      });
    },
    {
      concurrency: 5,
    },
  );
  return handlerResult;
}
