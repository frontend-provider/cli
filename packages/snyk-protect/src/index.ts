#!/usr/bin/env node

import protect from './lib';

async function main() {
  const projectPath = process.cwd();
  await protect(projectPath);
}

if (require.main === module) {
  main();
}
