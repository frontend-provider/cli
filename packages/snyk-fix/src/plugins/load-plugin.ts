import { UnsupportedTypeError } from '../lib/errors/unsupported-type-error';
import { pythonFix } from './python';
import { FixHandler } from './types';

export function loadPlugin(type: string): FixHandler {
  switch (type) {
    case 'pip': {
      return pythonFix;
    }
    case 'poetry': {
      return pythonFix;
    }
    default: {
      throw new UnsupportedTypeError(type);
    }
  }
}
