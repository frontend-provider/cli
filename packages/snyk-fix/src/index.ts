import * as debugLib from 'debug';
import { loadPlugin } from './plugins/load-plugin';
import { FixHandlerResult } from './plugins/types';

import { EntityToFix } from './types';

const debug = debugLib('snyk-fix:main');

export async function fix(
  entities: EntityToFix[],
): Promise<FixHandlerResult[]> {
  debug(`Requested to fix ${entities.length} projects.`);
  const entitiesPerType = groupEntitiesPerType(entities);
  const allResults: FixHandlerResult[] = [];
  // TODO: pMap this?
  for (const type of Object.keys(entitiesPerType)) {
    const handler = loadPlugin(type);
    const results = await handler(entitiesPerType[type]);
    allResults.push(results);
  }

  return allResults;
}

function groupEntitiesPerType(
  entities: EntityToFix[],
): {
  [type: string]: EntityToFix[];
} {
  const entitiesPerType: {
    [type: string]: EntityToFix[];
  } = {};
  for (const entity of entities) {
    const type = entity.scanResult.identity.type;
    if (entitiesPerType[type]) {
      entitiesPerType[type].push(entity);
      continue;
    }
    entitiesPerType[type] = [entity];
  }
  return entitiesPerType;
}
