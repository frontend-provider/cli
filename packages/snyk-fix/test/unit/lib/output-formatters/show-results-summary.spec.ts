import stripAnsi = require('strip-ansi');
import {
  CustomError,
  ERROR_CODES,
} from '../../../../src/lib/errors/custom-error';
import {
  generateOverallSummary,
  generateSuccessfulFixesSummary,
  generateUnresolvedSummary,
  formatIssueCountBySeverity,
  showResultsSummary,
} from '../../../../src/lib/output-formatters/show-results-summary';
import { FixHandlerResultByPlugin } from '../../../../src/plugins/types';
import { ErrorsByEcoSystem } from '../../../../src/types';
import { generateEntityToFix } from '../../../helpers/generate-entity-to-fix';

describe('generateOverallSummary', () => {
  it('has fixed & failed', async () => {
    const entity = generateEntityToFix(
      'pip',
      'requirements.txt',
      JSON.stringify({}),
    );
    const entityFailed = generateEntityToFix(
      'pip',
      'bad.txt',
      JSON.stringify({}),
    );
    const resultsByPlugin: FixHandlerResultByPlugin = {
      python: {
        succeeded: [
          {
            original: entity,
            changes: [
              {
                success: true,
                userMessage: 'Upgraded Django from 1.6.1 to 2.0.1',
                issueIds: ['vuln-2'],
              },
            ],
          },
        ],
        failed: [
          {
            original: entityFailed,
            error: new CustomError('Failed!', ERROR_CODES.MissingFileName),
          },
        ],
        skipped: [],
      },
    };
    const res = await generateOverallSummary(resultsByPlugin, {}, [], {});
    expect(stripAnsi(res.summary)).toMatchSnapshot();
  });

  it('has fixed only', async () => {
    const entity = generateEntityToFix(
      'pip',
      'requirements.txt',
      JSON.stringify({}),
    );
    const resultsByPlugin: FixHandlerResultByPlugin = {
      python: {
        succeeded: [
          {
            original: entity,
            changes: [
              {
                success: true,
                userMessage: 'Upgraded Django from 1.6.1 to 2.0.1',
                issueIds: ['vuln-1'],
              },
            ],
          },
        ],
        failed: [],
        skipped: [],
      },
    };
    const res = await generateOverallSummary(resultsByPlugin, {}, [], {});
    expect(stripAnsi(res.summary)).toMatchSnapshot();
    expect(res.count).toEqual(1);
  });

  it('has failed only', async () => {
    const entityFailed = generateEntityToFix(
      'npm',
      'package.json',
      JSON.stringify({}),
    );
    const resultsByPlugin: FixHandlerResultByPlugin = {
      python: {
        succeeded: [],
        failed: [
          {
            original: entityFailed,
            error: new CustomError('Failed!', ERROR_CODES.MissingFileName),
          },
        ],
        skipped: [],
      },
    };
    const res = await generateOverallSummary(resultsByPlugin, {}, [], {
      dryRun: true,
    });
    expect(stripAnsi(res.summary)).toMatchSnapshot();
    expect(res.count).toEqual(1);
  });

  it('has skipped & failed & plugin errors', async () => {
    const entity = generateEntityToFix(
      'pip',
      'requirements.txt',
      JSON.stringify({}),
    );
    const entityFailed = generateEntityToFix(
      'pip',
      'bad.txt',
      JSON.stringify({}),
    );
    const entitySkipped = generateEntityToFix(
      'pip',
      'Pipfile',
      JSON.stringify({}),
    );
    const resultsByPlugin: FixHandlerResultByPlugin = {
      python: {
        succeeded: [
          {
            original: entity,
            changes: [
              {
                success: true,
                userMessage: 'Upgraded Django from 1.6.1 to 2.0.1',
                issueIds: ['vuln-1'],
              },
            ],
          },
        ],
        failed: [
          {
            original: entityFailed,
            error: new CustomError('Failed!', ERROR_CODES.MissingFileName),
          },
        ],
        skipped: [
          { original: entitySkipped, userMessage: 'Pipfile is not supported' },
        ],
      },
    };
    const res = await generateOverallSummary(resultsByPlugin, {}, [], {
      dryRun: true,
    });
    expect(stripAnsi(res.summary)).toMatchSnapshot();
    expect(res.count).toEqual(3);
  });
  it('has fixed & failed & not vulnerable', async () => {
    const entity = generateEntityToFix(
      'pip',
      'requirements.txt',
      JSON.stringify({}),
    );
    const entityFailed = generateEntityToFix(
      'pip',
      'bad.txt',
      JSON.stringify({}),
    );
    const resultsByPlugin: FixHandlerResultByPlugin = {
      python: {
        succeeded: [
          {
            original: entity,
            changes: [
              {
                success: true,
                userMessage: 'Upgraded Django from 1.6.1 to 2.0.1',
                issueIds: ['vuln-2'],
              },
            ],
          },
        ],
        failed: [
          {
            original: entityFailed,
            error: new CustomError('Failed!', ERROR_CODES.MissingFileName),
          },
        ],
        skipped: [],
      },
    };
    const notVulnerable = generateEntityToFix(
      'pip',
      '',
      JSON.stringify({}),
      false,
    );
    const res = await generateOverallSummary(
      resultsByPlugin,
      {},
      [notVulnerable],
      {
        dryRun: true,
      },
    );
    expect(stripAnsi(res.summary)).toMatchSnapshot();
  });

  it('100% not vulnerable', async () => {
    const resultsByPlugin: FixHandlerResultByPlugin = {
      python: {
        succeeded: [],
        failed: [],
        skipped: [],
      },
    };
    const notVulnerable = generateEntityToFix(
      'pip',
      '',
      JSON.stringify({}),
      false,
    );
    const res = await generateOverallSummary(
      resultsByPlugin,
      {},
      [notVulnerable],
      {
        dryRun: true,
      },
    );
    expect(stripAnsi(res.summary)).toMatchSnapshot();
  });
});

describe('generateSuccessfulFixesSummary', () => {
  it('has fixed & failed', () => {
    const entity = generateEntityToFix(
      'pip',
      'requirements.txt',
      JSON.stringify({}),
    );
    const resultsByPlugin: FixHandlerResultByPlugin = {
      python: {
        succeeded: [
          {
            original: entity,
            changes: [
              {
                success: true,
                userMessage: 'Upgraded Django from 1.6.1 to 2.0.1',
                issueIds: ['vuln-2'],
              },
              {
                success: false,
                reason: 'Version not compatible',
                userMessage: 'Failed to upgrade transitive from 6.1.0 to 6.2.1',
                tip: 'Apply the changes manually',
                issueIds: ['vuln-1'],
              },
            ],
          },
        ],
        failed: [],
        skipped: [],
      },
    };
    const res = generateSuccessfulFixesSummary(resultsByPlugin);
    expect(stripAnsi(res)).toMatchSnapshot();
  });
});

describe('generateUnresolvedSummary', () => {
  it('has failed upgrades & unsupported', async () => {
    const entity = generateEntityToFix(
      'pip',
      'requirements.txt',
      JSON.stringify({}),
    );
    const entityFailed = generateEntityToFix(
      'npm',
      'package.json',
      JSON.stringify({}),
    );
    const resultsByPlugin: FixHandlerResultByPlugin = {
      python: {
        succeeded: [
          {
            original: entity,
            changes: [
              {
                success: true,
                userMessage: 'Upgraded Django from 1.6.1 to 2.0.1',
                issueIds: ['vuln-2'],
              },
              {
                success: false,
                reason: 'Version not compatible',
                userMessage: 'Failed to upgrade transitive from 6.1.0 to 6.2.1',
                tip: 'Apply the changes manually',
                issueIds: ['vuln-1'],
              },
            ],
          },
        ],
        failed: [],
        skipped: [],
      },
    };
    const exceptionsByScanType: ErrorsByEcoSystem = {
      python: {
        originals: [entityFailed],
        userMessage: 'npm is not supported',
      },
    };

    const res = await generateUnresolvedSummary(
      resultsByPlugin,
      exceptionsByScanType,
    );
    expect(stripAnsi(res.summary)).toMatchSnapshot();
    expect(res.count).toEqual(1);
  });
});

describe('formatIssueCountBySeverity', () => {
  it('all vuln severities present', async () => {
    const res = await formatIssueCountBySeverity({
      critical: 1,
      high: 3,
      medium: 15,
      low: 300,
    });
    expect(stripAnsi(res)).toMatchSnapshot();
  });

  it('all vuln severities absent', async () => {
    const res = await formatIssueCountBySeverity({});
    expect(stripAnsi(res)).toMatchSnapshot();
  });

  it('all vuln severities 0', async () => {
    const res = await formatIssueCountBySeverity({
      critical: 0,
      high: 0,
      medium: 0,
      low: 0,
    });
    expect(stripAnsi(res)).toMatchSnapshot();
  });

  it('Critical vulns 0', async () => {
    const res = await formatIssueCountBySeverity({
      critical: 0,
      high: 1,
      medium: 2,
      low: 60,
    });
    expect(stripAnsi(res)).toMatchSnapshot();
  });
});

describe('showResultsSummary', () => {
  it('has failed, skipped, successful & plugin errors', async () => {
    const entity = generateEntityToFix(
      'pip',
      'requirements.txt',
      JSON.stringify({}),
    );
    const entityNotSupported = generateEntityToFix(
      'npm',
      'package.json',
      JSON.stringify({}),
    );
    const entityFailed = generateEntityToFix('pip', '', JSON.stringify({}));
    const resultsByPlugin: FixHandlerResultByPlugin = {
      python: {
        succeeded: [
          {
            original: entity,
            changes: [
              {
                success: true,
                userMessage: 'Upgraded Django from 1.6.1 to 2.0.1',
                issueIds: ['vuln-1'],
              },
              {
                success: false,
                reason: 'Version not compatible',
                userMessage: 'Failed to upgrade transitive from 6.1.0 to 6.2.1',
                tip: 'Apply the changes manually',
                issueIds: ['vuln-2'],
              },
            ],
          },
        ],
        failed: [
          {
            original: entityFailed,
            error: new CustomError(
              'Missing required file name',
              ERROR_CODES.MissingFileName,
            ),
          },
        ],
        skipped: [],
      },
    };
    const exceptionsByScanType: ErrorsByEcoSystem = {
      python: {
        originals: [entityNotSupported],
        userMessage: 'npm is not supported',
      },
    };

    const res = await showResultsSummary(
      [],
      resultsByPlugin,
      exceptionsByScanType,
      { stripAnsi: true },
      3,
    );
    expect(res).toMatchSnapshot();
  });
  it('has unresolved only', async () => {
    const entityFailed = generateEntityToFix(
      'npm',
      'package.json',
      JSON.stringify({}),
    );
    const resultsByPlugin: FixHandlerResultByPlugin = {
      python: {
        succeeded: [],
        failed: [],
        skipped: [],
      },
    };
    const exceptionsByScanType: ErrorsByEcoSystem = {
      python: {
        originals: [entityFailed],
        userMessage: 'npm is not supported',
      },
    };

    const res = await showResultsSummary(
      [],
      resultsByPlugin,
      exceptionsByScanType,
      { stripAnsi: true },
      1,
    );
    expect(res).toMatchSnapshot();
  });
  it('called with nothing to fix', async () => {
    const resultsByPlugin: FixHandlerResultByPlugin = {
      python: {
        succeeded: [],
        failed: [],
        skipped: [],
      },
    };
    const exceptionsByScanType: ErrorsByEcoSystem = {};

    const res = await showResultsSummary(
      [],
      resultsByPlugin,
      exceptionsByScanType,
      { stripAnsi: true },
      0,
    );
    expect(res).toMatchSnapshot();
  });
  it('called with no vulnerable projects to fix', async () => {
    const resultsByPlugin: FixHandlerResultByPlugin = {
      python: {
        succeeded: [],
        failed: [],
        skipped: [],
      },
    };
    const exceptionsByScanType: ErrorsByEcoSystem = {};
    const notVulnerable = generateEntityToFix(
      'pip',
      '',
      JSON.stringify({}),
      false,
    );

    const res = await showResultsSummary(
      [notVulnerable],
      resultsByPlugin,
      exceptionsByScanType,
      {},
      1,
    );
    expect(stripAnsi(res)).toMatchSnapshot();
  });
});
