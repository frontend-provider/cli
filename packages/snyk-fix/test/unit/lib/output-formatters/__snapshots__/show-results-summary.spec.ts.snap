// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatIssueCountBySeverity Critical vulns 0 1`] = `"1 High | 2 Medium | 60 Low"`;

exports[`formatIssueCountBySeverity all vuln severities 0 1`] = `""`;

exports[`formatIssueCountBySeverity all vuln severities absent 1`] = `""`;

exports[`formatIssueCountBySeverity all vuln severities present 1`] = `"1 Critical | 3 High | 15 Medium | 300 Low"`;

exports[`generateOverallSummary 100% not vulnerable 1`] = `
"Summary:

  Command run in dry run mode. Fixes are not applied.
  1 items were not vulnerable

      "
`;

exports[`generateOverallSummary has failed only 1`] = `
"Summary:

  Command run in dry run mode. Fixes are not applied.
  1 items were not fixed

  1 issues: 1 High
  1 issues are fixable
  "
`;

exports[`generateOverallSummary has fixed & failed & not vulnerable 1`] = `
"Summary:

  Command run in dry run mode. Fixes are not applied.
  1 items were not fixed
  1 items were successfully fixed
  1 items were not vulnerable

  2 issues: 2 High
  2 issues are fixable
  1 issues were successfully fixed
"
`;

exports[`generateOverallSummary has fixed & failed 1`] = `
"Summary:

  1 items were not fixed
  1 items were successfully fixed

  2 issues: 2 High
  2 issues are fixable
  1 issues were successfully fixed
"
`;

exports[`generateOverallSummary has fixed only 1`] = `
"Summary:

  1 items were successfully fixed

  1 issues: 1 High
  1 issues are fixable
  1 issues were successfully fixed
"
`;

exports[`generateOverallSummary has skipped & failed & plugin errors 1`] = `
"Summary:

  Command run in dry run mode. Fixes are not applied.
  2 items were not fixed
  1 items were successfully fixed

  3 issues: 3 High
  3 issues are fixable
  1 issues were successfully fixed
"
`;

exports[`generateSuccessfulFixesSummary has fixed & failed 1`] = `
"Successful fixes:

  requirements.txt
  ✔ Upgraded Django from 1.6.1 to 2.0.1
  x Failed to upgrade transitive from 6.1.0 to 6.2.1
  Reason:  Version not compatible.
  Tip:     Apply the changes manually"
`;

exports[`generateUnresolvedSummary has failed upgrades & unsupported 1`] = `
"

Unresolved items:

  package.json
  ✖ npm is not supported"
`;

exports[`showResultsSummary called with no vulnerable projects to fix 1`] = `
"
✔ No vulnerable items to fix

Summary:

  1 items were not vulnerable

      "
`;

exports[`showResultsSummary called with nothing to fix 1`] = `
"
 ✖ No successful fixes"
`;

exports[`showResultsSummary has failed, skipped, successful & plugin errors 1`] = `
"
Successful fixes:

  requirements.txt
  ✔ Upgraded Django from 1.6.1 to 2.0.1
  x Failed to upgrade transitive from 6.1.0 to 6.2.1
  Reason:  Version not compatible.
  Tip:     Apply the changes manually

Unresolved items:

  pip project
  ✖ Missing required file name

  package.json
  ✖ npm is not supported

Summary:

  2 items were not fixed
  1 items were successfully fixed

  3 issues: 3 High
  3 issues are fixable
  1 issues were successfully fixed


Tip: Re-run in debug mode to see more information: DEBUG=*snyk* <COMMAND>. If the issue persists contact support@snyk.io"
`;

exports[`showResultsSummary has unresolved only 1`] = `
"
 ✖ No successful fixes


Unresolved items:

  package.json
  ✖ npm is not supported

Summary:

  1 items were not fixed

  1 issues: 1 High
  1 issues are fixable
  

Tip: Re-run in debug mode to see more information: DEBUG=*snyk* <COMMAND>. If the issue persists contact support@snyk.io"
`;
